cylc/flow/task_outputs.py

```diff
@@ -609,15 +624,23 @@ def _is_compvar_complete(self, compvar: str) -> Optional[bool]:
         else:
             raise KeyError(compvar)
 
-    def iter_required_messages(self) -> Iterator[str]:
+    def iter_required_messages(
+        self,
+        exclude=None
+    ) -> Iterator[str]:
         """Yield task messages that are required for this task to be complete.
 
         Note, in some cases tasks might not have any required messages,
         e.g. "completion = succeeded or failed".
+
+        Args:
+            exclude: Exclude one possible required messages, allowing
+            specification of all required outputs if succeeded or failed.
         """
         for compvar, is_optional in get_optional_outputs(
             self._completion_expression,
             set(self._message_to_compvar.values()),
+            force_optional=exclude
```

Comment:
Was it necessary to add this complexity to `get_optional_outputs`? Can it not be handled inside `skip.py`?

Reply:
I think it's reasonable there.